// angular2-polyfill/src/platform/bootstrap/directive.js
"use strict";
var utils = require('./utils');
function parseSelector(selector) {
    var regex = [
        // {key: 'E', value: /^([a-zA-Z])$/},
        { key: 'A', value: /^\[([a-zA-Z]+)\]$/ },
        { key: 'C', value: /^\.([a-zA-Z]+)$/ }
    ];
    for (var i = 0; i < regex.length; i++) {
        var result = selector.match(regex[i].value);
        if (result !== null) {
            return { restrict: regex[i].key, name: result[1] };
        }
    }
    ;
    throw new Error("Selector " + selector + " could not be parsed");
}
function bootstrap(ngModule, target) {
    var annotations = target.__annotations__;
    var directive = annotations.directive;
    var selector = parseSelector(directive.selector);
    var hostBindings = utils.parseHosts(directive.host || {});
    // Inject the services
    utils.inject(target);
    ngModule
        .controller(target.name, target)
        .directive(selector.name, [function () {
            var declaration = {
                restrict: selector.restrict,
                scope: {},
                bindToController: {},
                controller: target.name,
                controllerAs: '$ctrl',
                link: function (scope, el) { return utils.bindHostBindings(scope, el, hostBindings); }
            };
            // Bind inputs and outputs
            utils.bindInput(target, declaration);
            utils.bindOutput(target, declaration);
            return declaration;
        }]);
}
exports.bootstrap = bootstrap;
